src/k8s/pkg/k8sd/controllers/control_plane_configuration.go

```diff
@@ -70,14 +70,22 @@ func (c *ControlPlaneConfigurationController) reconcile(ctx context.Context, con
 	switch config.Datastore.GetType() {
 	case "external":
 		// certificates
-		if err := setup.EnsureExtDatastorePKI(c.snap, &pki.ExternalDatastorePKI{
+		anyUpdated, err := setup.EnsureExtDatastorePKI(c.snap, &pki.ExternalDatastorePKI{
 			DatastoreCACert:     config.Datastore.GetExternalCACert(),
 			DatastoreClientCert: config.Datastore.GetExternalClientCert(),
 			DatastoreClientKey:  config.Datastore.GetExternalClientKey(),
-		}); err != nil {
+		})
+
+		if err != nil {
 			return fmt.Errorf("failed to reconcile external datastore certificates: %w", err)
 		}
 
+		if anyUpdated {
+			if err := c.snap.RestartService(ctx, "kube-apiserver"); err != nil {
+				return fmt.Errorf("failed to restart kube-apiserver to apply configuration: %w", err)
+			}
+		}
```

Comment:
note: this might result in multiple kube-apiserver restarts. e.g. if both certificates and arguments have changed. this is not useful, because it might result in failures (e.g. restarting the server without the proper arguments) or extended service downtime.

instead, the service should only be restarted after all configs are in place, see the code suggestions

Reply:
Nice, good observation